packages/babel-parser/src/parser/expression.js

```diff
@@ -2725,4 +2733,163 @@ export default class ExpressionParser extends LValParser {
+    const node = this.startNode<N.MatchClause>();
+    const type = this.state.type;
+    this.next(); // skip "if" / "else"
+    if (type === tt._if) {
+      node.guard = this.parseHeaderExpression();
+    }
+    node.consequent = this.parseStatement("if");
+    return this.finishNode(node, "MatchClause");
+  }
+
+  parseWhenMatchClause(): N.MatchClause {
+    const node = this.startNode<N.MatchClause>();
+    this.next(); // skip "when"
+    this.expect(tt.parenL);
+    node.test = this.parseMaybeBinaryMatchPattern();
+    this.expect(tt.parenR);
+
+    if (this.isLineTerminator()) {
+      this.unexpected(
+        this.state.lastTokEnd,
+        Errors.LineTerminatorAfterMatchPattern,
+      );
+    }
+    node.consequent = this.parseStatement("do");
+    return this.finishNode(node, "MatchClause");
+  }
+
+  parseMatchPattern(): N.MatchPattern {
+    switch (this.state.type) {
+      case tt.slash:
+      case tt.slashAssign:
+        this.readRegexp();
+        return this.parseRegExpLiteral(this.state.value);
+      case tt.num:
+        return this.parseNumericLiteral(this.state.value);
+      case tt.bigint:
+        return this.parseBigIntLiteral(this.state.value);
+      case tt.string:
+        return this.parseStringLiteral(this.state.value);
+      case tt._null:
+        return this.parseNullLiteral();
+      case tt._true:
+        return this.parseBooleanLiteral(true);
+      case tt._false:
+        return this.parseBooleanLiteral(false);
+      case tt.name:
+        return this.parseIdentifier();
+      case tt.braceL:
+        return this.parseObjectMatchPattern();
+      case tt.bracketL:
+        return this.parseArrayMatchPattern();
+      default:
+        return;
+    }
+  }
+
+  parseMaybeBinaryMatchPattern(): N.BinaryMatchPattern | N.MatchPattern {
+    const node = this.startNode<N.BinaryMatchPattern>();
+    const lhs = this.parseMatchPattern();
+    if (this.match(tt.bitwiseOR) || this.match(tt.bitwiseAND)) {
+      node.left = lhs;
+      node.operator = this.match(tt.bitwiseOR) ? "or" : "and";
+      this.next();
+      node.right = this.parseMaybeBinaryMatchPattern();
+      return this.finishNode(node, "BinaryMatchPattern");
+    } else {
+      return lhs;
+    }
+  }
+
+  parseObjectMatchPattern(): N.ObjectMatchPattern {
+    const node = this.startNode<N.ObjectMatchPattern>();
+    this.expect(tt.braceL);
+    const properties: (N.AssignmentMatchProperty | N.RestMatchElement)[] = [];
+    while (!this.eat(tt.braceR)) {
+      if (this.match(tt.ellipsis)) {
+        properties.push(this.parseRestMatchElement());
+        this.checkCommaAfterRest(charCodes.rightCurlyBrace);
+        this.expect(tt.braceR);
+        break;
+      }
+      const node = this.startNode<N.AssignmentMatchProperty>();
+      node.kind = "init";
+      node.method = false;
+      this.parsePropertyName(node, /* isPrivateNameAllowed */ false);
+      if (this.eat(tt.colon)) {
+        node.value = this.parseMaybeBinaryMatchPattern();
```

Comment:
```suggestion
        node.value = this.parseMatchPattern();
```

ObjectMatchPattern can certainly be nested.